cachi2/core/package_managers/gomod.py

```diff
@@ -559,30 +758,62 @@ def _resolve_gomod(
 
     run_params = {"env": env, "cwd": app_dir}
 
+    go = Go()
+    go1_21 = version.Version("1.21")
+    go_base_version = go.version
+    go_mod_version_msg = "go.mod recommends/requires Go version: {}"
+
+    if (_ := _get_gomod_version(app_dir)) is None:
+        # Go added the 'go' directive to go.mod in 1.12 [1]. If missing, 1.16 is assumed [2].
+        # For our version comparison purposes we set the version explicitly to 1.20 if missing.
+        # [1] https://go.dev/doc/go1.12#modules
+        # [2] https://go.dev/ref/mod#go-mod-file-go
+        _ = "1.20"
+        go_mod_version_msg += " " + "(cachi2 enforced)"
+
+    go_mod_version = version.Version(_)
+
+    log.info(go_mod_version_msg.format(go_mod_version))
+
+    if go_mod_version >= go1_21 and go_base_version < go1_21:
+        # our base Go installation is too old and we need a newer one to support new keywords
+        go = Go(release="go1.21.0")
+    elif go_mod_version < go1_21 and go_base_version >= go1_21:
```

Comment:
And in case `gomod_version >= go1_21 and go_base_version >= go_121`, we can just leave Go handle any specific declared toolchain, right?

I mean, imagine go_base_version is `go1.21.1` and gomod_version is `go1.21.6`. Would Go download the specified toolchain for commands such as `download`?

Reply:
> And in case `gomod_version >= go1_21 and go_base_version >= go_121`, we can just leave Go handle any specific declared toolchain, right?
> 

Yes, cachi2 is just a middle man.  We can't be responsible for someone specifying `go 1.21.6` in their mod file and expecting the build will always pass as that's awfully specific and would require essentially everyone to download that specific toolchain before proceeding. My take on this is that it was never meant to work this way, it's just starting 1.21 Go now recognizes "micro" releases in version strings due to the `toolchain` directive. The `go` directive should only ever contain language version (looking at you Syft --> `1.21.0`) which guarantees that any toolchain build conforming to the 1.21 spec will build the project.

> I mean, imagine go_base_version is `go1.21.1` and gomod_version is `go1.21.6`. Would Go download the specified toolchain for commands such as `download`?

Not with `GOTOOLCHAIN=local`, in this case it'll fail with `go: go.mod requires go >= 1.21.6 (running go 1.21.4; GOTOOLCHAIN=local)`. With `GOTOOLCHAIN=auto` (future full support) will happily pass and download both the toolchain and dependencies.

Reply:
Note that I could make the clause more generic with `if go_mod_version > go_base_version` but that would essentially always download a toolchain during runtime which we likely don't want. Like I said, if anyone puts a specific toolchain version to the `go` directive, they should suffer the consequences because literally any CI would choke on that.

Reply:
Humm, whereas I agree with the approach, I'm just worried because some users are asking for Go 1.21 support specifically because some projects do have the `toolchain` directive. So this would make these projects fail if the specified toolchain does not match the one being used by Cachi2?

Reply:
I'm not sure I  follow your point, `toolchain` is only informative, not required. Adn in fact, bundled versions of Go 1.21 (e.g. Fedora) ship with `GOTOOLCHAIN=local` which means that `toolchain` is going to be ignored anyway. The whole point of this preliminary support is to make it consistent by enforcing GOTOOLCHAIN so that people don't use environments, where that bundled toolchain ships with `GOTOOLCHAIN=auto`. Can you elaborate more on your point?

Reply:
Sorry, I don't have all the details on how this works in my mind, so I might have gotten something wrong.

> Not with GOTOOLCHAIN=local, in this case it'll fail with go: go.mod requires go >= 1.21.6 (running go 1.21.4; GOTOOLCHAIN=local). 

I understood that, considering the above, the prefetch would fail if the `go.mod` file specifies a `toolchain` that is higher from the one being used. If that is not the case, then we're safe.

Reply:
But if the project specifies `go 1.22` and your base go is 1.21.x, it fails immediately no? (If you set `GOTOOLCHAIN=local` that is)

Previously it would have worked to a limited extent, the older Go just wouldn't necessarily support the features that the user might be trying to use.

With full toolchain suppor, cachi2 will download the right version of Go? But won't necessarily represent everything properly in the SBOM (unknown new features)

Reply:
```bash
$ go mod download
go: go.mod requires go >= 1.22 (running go 1.21.6; GOTOOLCHAIN=local)
```

Reply:
> But if the project specifies `go 1.22` and your base go is 1.21.x, it fails immediately no? (If you set `GOTOOLCHAIN=local` that is)
Previously it would have worked to a limited extent, the older Go just wouldn't necessarily support the features that the user might be trying to use.

That's right, but there's no 1.22 yet and by then, we should have full support merged.

So what exactly are we discussing here really? Let's summarize:

- The base (default) 1.21 enforces usage of 1.21 as the least common denominator if Go finds any of the dependencies to specify 1.21 while your project's version < 1.21 
    - --> dirty git tree --> we fall back to 1.20
- The base version is < 1.21 (on local environments) while the project specifies >= 1.21
   - 1.20 may or may not be able to build the project, so compared to the standard behaviour where it could fail, we go the extra mile to download and install 1.21 to eliminate the potential failure
 
So, is this whole discussion about cachi2 not going the extra mile of downloading literally ANY toolchain when `base>=1.21 and gomod>base` compared to the other cases?
The question then is, is it worth adding that functionality if we already know we'll have revert it soon when we start setting `GOTOOLCHAIN=auto` and hence render that case irrelevant?

Reply:
I think it's only a question of priority for the full toolchain support. These issues make it a higher priority than originally expected IMO